### PR TITLE
feat(store): provide better TS errors for action creator props

### LINKED ISSUE #2892: props doesn't accept HttpErrorResponse
## Minimal reproduction of the bug/regression with instructions:
I want to pass the HttpError through an action and so I do the following:

```ts
export const someFail = createAction(
  '[Feature] Some fail',
  props<HttpErrorResponse>(),
);
```

... however it doesn't compile:

Captura de Tela 2021-01-21 às 22 03 14
Captura de Tela 2021-01-21 às 22 03 51

## Expected behavior:
`props` accept `HttpErrorResponse`.

## Versions of NgRx, Angular, Node, affected browser(s) and operating system(s):
Ngrx 10, Angular 11.

## I would be willing to submit a PR to fix this issue

[x] Yes (Assistance is provided if you need help submitting a pull request)
[ ] No

### PR DESCRIPTION
**Pull request by david-shortman**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Closes #2892

## Does this PR introduce a breaking change?

```
[x] Yes - usage of `props` outside of an action creator now breaks for invalid types
[ ] No
```

## Other information

Thanks to the advice from [Nadia Chibrikova on StackOverflow](https://stackoverflow.com/a/68080083/11087018), `props` can have its generic argument asserted to pass `NotAllowedCheck`:

Screen Shot 2021-06-30 at 21 43 28

This provides a more precise error than the current one which is hidden within TS outputting multiple errors for a mismatch for each overload of `createAction`:

Screen Shot 2021-06-30 at 21 44 48

Additionally, I've updated the error messages to be hopefully more explicit about what is wrong.

UPDATE: With a suggestive playground link, I took @markostanimirovic's subtle suggestion to explicitly restrict primitive types.

I've also separated the error messages to be more contextually meaningful when using `props` vs when providing a creator function to `createAction`.

### CODE REVIEW

**Review by timdeschryver — APPROVED** [review]
Neat!

File: modules/store/src/action_creator.ts
Comment (timdeschryver) [inline modules/store/src/action_creator.ts:128]: An afterthought...
Could we extract this into a PropsNotAllowed type?
Similar to the [NotAllowedCheck](https://github.com/ngrx/platform/blob/master/modules/store/src/models.ts#L83) type.

File: modules/store/src/action_creator.ts
Comment (markostanimirovic) [inline modules/store/src/action_creator.ts:134]: Hey @david-shortman , nice idea 🙂

Try this: [playground Link](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAYwgOwM7wIZShgnqgQSmADkIYCAbSiAd2ABMBZVAczgF44ByDBGAJYpExDDGhwwUCGFSIMyZOTgAjYHAUbsebgG4AUDFxh1RHPiKlyVGvQac4RkxABmW84WJkK1OoxasBvqgkLCIKOiOxsAACtImsLgAkqjeNn7MbA68-ELIIsBiElIycggKSvAAFhgAbuoYkvHAiXDIGAC2jHAABk7APXqG0XAAKtFxMi1GKWm+dg79rlEmkwkzqdbz-mxBIdDwSGjwwB1gRgDyKgBWwPyeVj62O+xcOYLCCKLiUE2l8oplGoNPlTudcHAIDc7jAhv04ABRM6XaH3SxzZ72LhLNxglG3NFeLaYgJ7cAHcLHJoCDoCQT1B4YjIBbJ8D75L6FH5-WQAyqqBrU2n09S1DCUACuwDhIziNLpAgZ6OJGUW0WWUnlIsZKrspP0wxMcDlwsV6i46CgAmQ7AAPm0JR01L97SoBKxrfBXRAIJRCvl7cgJdQ4PaJcgGMAXNbGGTQvB4Uy7ABhKp3ADWAB5RnBQDBgBG5FCCTAAHwOHN5gsMOQKXAAbQAuvo4HAAPxwMx4HVPDItuAALjGuZA+cLcAA3itgEO63AAL79jvjVbNRKzXWMftD9PAXDLSuj6tyZDAepQJeI5G4K4lnvpOzbuDh9NKWjIOMUxObhhrVCpjNs3LDh+0PMca0hVEYH7VsOyTRgAIQLNRlLGDW0HYcq3HE0FXqND0PbY0rVNJUiV7R8CIIocXzfD8DX2MIXHDXJhBKWRM37P8R3AuQ-0Q9MABpOPif80yQhx4N-ET+MzP9UNLAAKABKSd+2IGAJSgfIJ3nDRa2QXADEXfQ2NQTMp36IdLWtQI4BAIcgydFoF0UpSglMzNrJtJtXPckTzPnXyDQ8xznV8oA) 

```suggestion
export function props<
  Props extends PropsCheck,
  PropsCheck = NotAllowedPropsCheck<Props>
>(): ActionCreatorProps<Props> {
```

File: modules/store/src/models.ts
Comment (david-shortman) [reply modules/store/src/models.ts:98]: I wonder if there's a reason Typescript doesn't provide this type already...

Comment (david-shortman) [reply modules/store/src/models.ts:98]: Their site says

> object is a type that represents the non-primitive type, i.e. anything that is not number, string, boolean, bigint, symbol, null, or undefined.

So it seems like checking if something `extends` object would be sufficient.

Perhaps I shouldn't have an explicit check for primitives then.

Comment (david-shortman) [reply modules/store/src/models.ts]: I need an explicit check because `unknown` does not extend `object`, and so it must be differentiated (but we can't check `extends unknown` because every type is assignable to `unknown`).

File: modules/store/src/models.ts
Comment (markostanimirovic) [inline modules/store/src/models.ts:2]:
```suggestion
```

**Review by markostanimirovic — CHANGES_REQUESTED** [review]
It looks much better now 🎉

**Review by markostanimirovic — APPROVED** [review]
Great work David! Thank you 🥇 

There are no breaking changes when `props` is used within `createAction`, but `props` itself has a breaking change. So, I would wait until v13 to merge this improvement.

**Comment by david-shortman** [comment]
> There are no breaking changes when `props` is used within `createAction`, but `props` itself has a breaking change. So, I would wait until v13 to merge this improvement.

Oh yeah, good point!

**Review by timdeschryver — APPROVED** [review]
LGTM 👍

**Pull request merged**
